[Wait for gst] Video: Fix MPEG-TS UDP streaming / Fix Android Recording - fix crashes (#9830, #9838)

fredowski opened the pull request:
### Problem
Decodebin3 breaks the MPEG-TS/UDP streaming. Decodebin3 is still named "experimental" on the gstreamer site. So this PR fixes the following problems:

Closes: MPEG-TS not streaming on Android #9830, #10200 
Closes: Recording not working on Android #9838

### What is done

This patch reverts to decodebin.  I reverted including the previously working recording fix for android. Reverting to decodebin makes threading problems transparent which were not visible in decodebin3 and which were introduced by the new videoreceiver design. The root problem is the addition of gstreamer elements in the callbacks of gstreamer. The callbacks run on threads of the existing gstreamer elements. This becomes apparent as "flickering" problems on Android and MacOS. This can be fixed by moving the gstreamer element addition back to the main gstreamer thread. This closes the problems described  #10396 and #10380.

The new videoreceiver design relies on dynamic pipeline modifications also for the video display part. I think the concept of first starting the pipeline with "start" and then adding the decoder part with "startdecoding" on an already running pipeline made threading problems visible. Depending on the platform, different numbers of threads are active. There is the main loop, then the "gstreamer main thread = VideoReceiver", then the "Rendering thread" on Android and then the threads which are introduced by gstreamer itself depending on the pipeline.  After starting and stopping a stream three times, the whole qgc application freezes on Android. I could fix that by backporting bugfixes from the qt gstreamer-plugins release 1.18.6 to qgroundcontrol gstreamer-plugins. I only backported bugfixes and not the additional overlay function which requires a complete switch to 1.18.6. Maybe it would be safer to step back and configure the video streaming pipeline including the decoding branch at start. Then only the recording branch would be added dynamically to the pipeline. But that is not part of this PR.

**This PR has to wait on https://github.com/mavlink/gst-plugins-good/pull/2 and https://github.com/mavlink/qgroundcontrol/pull/10396**

and then this PR has to be updated to use the gstreamer plugins from the mavlink repository.

### Tests

I tested this PR on MacOS, Android and Windows 11 and I could not find problems. On MacOS I tested this on top of PR #10396.

 I tested if the video is displayed. For a subset of resolutions I tested if the recording works.

**UDP H.264 results**

`gst-launch-1.0 videotestsrc ! 'video/x-raw, width=720, height=576' ! x264enc ! 'video/x-h264, profile=main' ! rtph264pay ! udpsink host=192.168.1.165 port=5600` 

I tested the following other resolutions

* 1920 x 1080 => worked, recording tested and works
* 720 x 576 => worked
* 640 x 480 => worked
* 320 x 240 => worked

**MPEG-TS (h.264) results**

I streamed a video from VLC Player on a Mac via MPEG-TS. I disabled the audio channel. This worked. This is a screenshot: 

![Screenshots 9_16_2022 4_04_24 PM](https://user-images.githubusercontent.com/112891091/190659278-22718dcc-61de-42cf-829d-7a6c2082c984.png)

I generated testdata on a Mac and streamed it to the windows computer. The testdata was generated with gstreamer 1.18.6. This is the script I used:

`gst-launch-1.0 videotestsrc ! 'video/x-raw, width=720, height=576' ! x264enc ! 'video/x-h264, profile=main' ! mpegtsmux ! queue ! udpsink host=192.168.1.165 port=5600` => worked

I tested the following resolutions

* 1920 x 1080 => worked, recording tested and works
* 720 x 576 => worked
* 640 x 480 => worked
* 320 x 240 => worked


@andrewvoznytsa: Can you maybe review this PR?

booo commented on .gitmodules:
```diff
@@ -9,7 +9,7 @@
 [submodule "libs/qmlglsink/gst-plugins-good"]
 	path = libs/qmlglsink/gst-plugins-good
-	url = https://github.com/mavlink/gst-plugins-good.git
+	url = https://github.com/fredowski/gst-plugins-good.git
```
You should remove this commit once your changes are merged upstream.

fredowski replied:
That commit must not be merged to master. This PR has to wait on https://github.com/mavlink/gst-plugins-good/pull/2 . That are the backported bugfixes for gst-plugins.

fredowski commented:
As discussed today during the QGC devcall with @zdanek and @mrpollo a test on a real linux machine needs to be done with the above described formats.

zdanek commented:
@fredowski I tested it on my Ubuntu 20.04. I tested:
* UDP 264
* MPEGTS 264
* RTSP 264 with quite decent Topotek gimball eth camera with thermal video (see in the right corner :)

Also I tested H265 UDP but there are some problems reported by GST. I enabled GST and video debug:
` GST_DEBUG=3 ./staging/QGroundControl  --logging:VideoAllLog`

Two H265 sources - FFMPEG and GST:
```

```
ffmpeg -f v4l2 -i /dev/video0 -pix_fmt yuv420p -c:v libx265 -preset ultrafast -x265-params crf=23 -strict experimental -f rtp udp://192.168.0.10:5600
```
gst-launch-1.0 videotestsrc ! 'video/x-raw, width=1024, height=768' ! x265enc ! 'video/x-h265, profile=main' ! rtph265pay ! udpsink host=192.168.0.10 port=5600
```

with both video does work  but with huge amount of FIXMEs and WARNINGs emited from GST  we need to check / decide if it can be fixed somehow. This needs to investigated. Maybe `queue` or `videoconvert` elements in the pipeline? During my experiments with gst I often had to put those elements into pipeline to get video work. :thinking: 

Below is a small piece of logs while playing h265 from both mentioned sources
```
0:00:21.856096351 157247 0x7f5d3c001800 WARN            videodecoder gstvideodecoder.c:2761:gst_video_decoder_prepare_finish_frame:<avdec_h265-0> decreasing timestamp (0:00:17.035964457 < 0:00:17.038971460)
0:00:21.862467736 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862552799 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862599684 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862645775 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862689366 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862738779 157247 0x7f5d44002360 FIXME           rtph265depay gstrtph265depay.c:1310:gst_rtp_h265_depay_process:<rtph265depay1> Assuming DONL field is not present
0:00:21.862897925 157247 0x7f5d44002360 FIXME              h265parse gsth265parse.c:1850:gst_h265_parse_parse_frame:<h265parse1> Implement timestamp/duration interpolation based on SEI message
0
```

zdanek commented:
@fredowski sorry, I was writing this epic for a dozen of minutes and forgot about nice screenshots :)
Regarding H265 I do agree that this can be fixed (if so) in other PR. Most popular H264 works great.

I have only builtin Intel graphics card. It supports OpenGL and some basics but it's not a something great :)

Screens:
h265
![Screenshot from 2022-10-10 12-39-26](https://user-images.githubusercontent.com/440614/194855650-4a89ef89-06c6-437d-88a1-c0afe3d4c1a6.png)

rtsp h264
![Screenshot from 2022-10-10 11-09-31](https://user-images.githubusercontent.com/440614/194855685-93535537-b467-4fa4-b7f3-9835acc6de4a.png)

udp h264 with lower res (as it can be seen it's blurry as it's upscaled)
![Screenshot from 2022-10-10 11-04-39](https://user-images.githubusercontent.com/440614/194855737-e342dbb7-6c04-4b6a-8c7f-14f04b5caa78.png)

Hot plugging works great. I mean - enabling video in QGC and starting and stopping and again of video source works good.

fredowski commented:
Very nice!

fredowski commented:
I rebased this to include the RTSP fix from the merged PR #10409.

jensensi commented:
Hello @fredowski 

I have now tried the APK [https://github.com/mavlink/qgroundcontrol/actions/runs/3237541415](https://github.com/mavlink/qgroundcontrol/actions/runs/3237541415) provided by fredowski for PR [#10420](https://github.com/mavlink/qgroundcontrol/pull/10420) on my device to check if the display and recording works?

I have only tested the PTSP interface, but using both the 'Bunny' video and also with my IP Camera. In this post I provide the findings from these test with the provided APK.

**System Information**
Operating System: [Android 9, FPV SIYI MK15 ]
Qgroundcontrol: [QGroundControl64.apk](https://github.com/mavlink/qgroundcontrol/actions/runs/3237541415) provided by fredowski

**Test**
_1) Source: "RTSP Video Stream". File format: "mkv" (url = Bunny). Video decode priority: "Default"._
One of two results are experienced when application is launched:

- The video display show "WAITING FOR VIDEO". No video is displayed. The recording video button is enabled (red/round).
Nothing happens when the record button is pressed, i.e. the button is unchanged (red/round).
The video folder contain a new video file, which is empty (0 bytes).

- The video display show the Bunny video. The recording video button is enabled (red/round).
When the record button is pressed, the button change shape from round to square (recording started) and remains square until it is pressed again (recording stopped).
The video folder contain a new video file, which is a recording of the 'Bunny' video stream.
(I have not experienced the 'flickering gui elements' as described in #10396 )

_2) Source: "RTSP Video Stream". File format: "mkv" (url = IP Camera). Video decode priority: "Default"._
Consistent result:

- The video display show "WAITING FOR VIDEO". No video is displayed. The recording video button is enabled (red/round).
When the record button is pressed, the button change shape from round to square (recording started) as in the successful case of the Bunny video, but it remains unchanged (square/red) when the button is pressed again.
The video folder contain a new video file, which is empty (0 bytes).

**Log files**
I have included log files for the two possible outcomes of test 1) 'Bunny' and also for test 2) 'Camera':
I have created the log files using similar settings as @fredowski:
- VideoAllLog = enabled
- GStreamer Debug = FixMe (GST_DEBUG=3)

_Test 1) - Source: Bunny video - Result: WAITING FOR VIDEO_
[Bunny-nok.txt](https://github.com/mavlink/qgroundcontrol/files/9774950/Bunny-nok.txt)

_Test 1) - Source: Bunny video - Result: Bunny video (and recording possible)_
[Bunny-ok.txt](https://github.com/mavlink/qgroundcontrol/files/9774953/Bunny-ok.txt)

_Test 2) - Source: IP Camera - Result: WAITING FOR VIDEO_
[Camera-nok.txt](https://github.com/mavlink/qgroundcontrol/files/9774954/Camera-nok.txt)

_Note! 
I have not yet had the chance to review the log file for my own camera in detail at this time, but I do notice that it is quite different from the Bunny case._

fredowski commented:
Thanks for testing! The problems for the the bunny video and the camera seem different. For your camera I have the impression that the stream format is not supported by gstreamer.

```
[E] at ../gst/videoparsers/gsth264parse.c:3450 - "<h264parse1> H.264 AVC caps, but no codec_data"
[E] at ../gst/videoparsers/gsth264parse.c:3476 - "<h264parse1> refused caps video/x-h264, stream-format=(string)avc, alignment=(string)au, width=(int)640, height=(int)352, framerate=(fraction)0/1, chroma-format=(string)4:2:0, bit-depth-luma=(uint)8, bit-depth-chroma=(uint)8, colorimetry=(string)1:3:5:1, parsed=(boolean)true, profile=(string)high, level=(string)3"
```

Does the bunny video show up eventually if you just wait? Or does it never display? The stream should tear down if no video is available after 8 seconds? Does that happen, i.e. do you see repeating pattern in the log? Does it display if you switch the video source to "None" and then back to "RTSP"? 

Can you display the Camera with the build for another platform like linux or windows?

Can you display the camera with a gstreamer pipeline via gst-launch ?

jensensi commented:
Hello @fredowski,

> Does the bunny video show up eventually if you just wait? Or does it never display?

No. I can wait for several minutes and nothing happen. The 'WAITING FOR VIDEO' message remains.

> The stream should tear down if no video is available after 8 seconds? Does that happen, i.e. do you see repeating pattern in the log? 

I have included a log with a duration of +30 seconds.

> Does it display if you switch the video source to "None" and then back to "RTSP"?

No.

> Can you display the Camera with the build for another platform like linux or windows?

Yes, it normally works on my Windows 10 PC that I use for development/build.

But, with the test setup used above, IT NO LONGER WORKS??
Why?? For the above test, I changed the camera output to H264 (to match the Bunny video).

_Changing the Camera output to H265_
If I change the Camera output back to H265, my Windows 10 PC works again - and the video is also displayed on your Android build. So, I have created new log files for the H265 camera setting.

> Can you display the camera with a gstreamer pipeline via gst-launch ?

I have not tried this method yet.


**Log files**

_Test: - Platform: Android - Source: Bunny video - Result: WAITING FOR VIDEO (wait for 30+ seconds)_
A longer log (30+ seconds) from the situation where the display maintain WAITING FOR VIDEO
[Bunny-nok-30sec.txt](https://github.com/mavlink/qgroundcontrol/files/9776392/Bunny-nok-30sec.txt)

_Test:  -Platform: Android - Source: IP Camera (H265) - Result: Video feed_
The camera feed is displayed. Nothing happens when pressing the record button.
[Camera-no-rec (android).txt](https://github.com/mavlink/qgroundcontrol/files/9776439/Camera-no-rec.android.txt)

_Test:  -Platform: Windows 10 - Source: IP Camera (H265) - Result: Video feed_
The camera feed is displayed. Recording starts/stops when pressing the record button.
[Camera-ok-rec (win10).txt](https://github.com/mavlink/qgroundcontrol/files/9776460/Camera-ok-rec.win10.txt)

zdanek commented:
@jensensi 
you've used `Rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4`
can you please use small-cap `rtsp` instead `Rtsp`? Silly but it might change something.
`rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4`
and post any results?

jensensi commented:
@zdanek 
Yes, you never know :-)
Here are two logs using lowercase 'rtsp' instead of 'Rtsp'.

- VideoAllLog = enabled
- GStreamer Debug = FixMe (GST_DEBUG=3)

_1) Bunny video shown in display when QGC is started_
[Bunny-ok.txt](https://github.com/mavlink/qgroundcontrol/files/9791912/Bunny-ok.txt)

_2) WAITING FOR VIDEO in display when QGC is started (some minutes)_
[Bunny-waiting-long.txt](https://github.com/mavlink/qgroundcontrol/files/9791911/Bunny-waiting-long.txt)

Note! I experience 'WAITING FOR VIDEO' in 1 of 5-8 cases, so most of the time it works well :-)

fredowski commented:
@jensensi: I added a check to ignore audio streams. If the media stream contains video and audio, then it depends on timing if the audio or the video data is presented first on a source pad. The first pad that is created on a source is linked. If the first pad provides audio instead of video, then the further processing doese not work. I added a check to ignore non-video streams. I tested this with the rtsp example bunny stream:

`Rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4`

with video source "RTSP Video Stream". Then I tested the MPEG-TS with a h264 encoded video containing also audio with the VLC player. The QGC video source is then set to "MPEG-TS (h.264) Video Stream". 

With video source set to "MPEG-TS (h.264) Video Stream", I can also see the "Tears of Steele" h.265 movie from here: https://h265.webmfiles.org streamed via VLC in MPEG-TS format. So the MPEG-TS setting in QGC is not restricted to h.264 - at least on MacOS. 

Can you maybe try this on your Android device?

fredowski commented:
![Bildschirmfoto 2022-10-18 um 15 05 45](https://user-images.githubusercontent.com/1280457/196438104-8b70f29c-cc5c-4a8f-8904-7578fe308ac2.jpg)

"Tears of Steel" in h.265 on QGC on MacOS... The Android build with the video stream check is here: 

https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680

Thanks for your detailed reports!

zdanek commented:
@fredowski who uses QGC to watch great _Tears of steel_ :rofl: 
I'm glad that is works :smile:

jensensi commented:
@fredowski
I have tested the new build [https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) on my Android device, using the Bunny video [rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4](rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4) as RTSP streaming source.

I have restarted the application +20 consecutive times, while anticipating that the app view would display "WAITING FOR VIDEO". But it doesn't?? 

In every of the +20 startup of the application, the Bunny video is displayed perfectly. :-)
I will say that the "WAITING FOR VIDEO" situation no longer occurs with the Bunny video on my device!!

THANK YOU!! WELL DONE!!

It gives me new hope that I may also be able to display my camera stream in H.264 mode on this software? 
I will test and return with updates :-)

jensensi commented:
@fredowski
I have extended the Bunny test [above](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282749541) to also include RTSP streaming from my IP camera. 

**Testing**
I have tested the on 2 software builds for Android:

- The [Bunnybuild](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) your latest build that disregard non-video streams for the first pad
- The [Standardbuild](https://qgroundcontrol.s3-us-west-2.amazonaws.com/latest/QGroundControl32.apk) QGroundControl v.4.2.3 64 bit (latest build 2022.10.22)

I test with the camera streaming H264 (mainprofile) video and inspired by your recent [findings](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282352495) I have tested both WITH and WITHOUT audio data in the stream.

The streaming settings are configured as defined by the test case, before the app is started.
The logs are created using the settings:
- VideoAllLog = enabled
- GStreamer Debug = FixMe (GST_DEBUG=3)


**Findings**

_Test 1) Software: [Bunnybuild](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) | Video source: IP camera, H.264 video stream_ 

- Test case 1.1) Bunnybuild - WITH audio stream
The video display show "WAITING FOR VIDEO". No video is displayed. The recording video button is enabled (red/round).
When the record button is pressed, the button change shape from round to square (recording started), but it remains unchanged (square/red) when the button is pressed again. The video folder contain a new video file, which is empty (0 bytes).
Logfile: [Bunnybuild-camera-h264-audio.txt](https://github.com/mavlink/qgroundcontrol/files/9814071/Bunnybuild-camera-h264-audio.txt)

- Test case 1.2) Bunnybuild - WITHOUT audio stream
The video display show "WAITING FOR VIDEO". No video is displayed. The recording video button is enabled (red/round).
When the record button is pressed, the button change shape from round to square (recording started), but it remains unchanged (square/red) when the button is pressed again. The video folder contain a new video file, which is empty (0 bytes).
Logfile: [Bunnybuild-camera-h264-noaudio.txt](https://github.com/mavlink/qgroundcontrol/files/9814073/Bunnybuild-camera-h264-noaudio.txt)

_Test 2) Software:  [Standardbuild](https://qgroundcontrol.s3-us-west-2.amazonaws.com/latest/QGroundControl32.apk) | Video source: IP camera, H.264 video stream_ 

- Test case 2.1) Standardbuild - WITH audio stream
The video display show the video stream from my IP camera.  The recording video button is enabled (red/round). Nothing happens when the record button is pressed, i.e. the button is unchanged (red/round). The video folder is empty afterwards.
Logfile: [QGC-v4.2.3-camera-h264-audio.txt](https://github.com/mavlink/qgroundcontrol/files/9814074/QGC-v4.2.3-camera-h264-audio.txt)

- Test case 2.2) Standardbuild - WITHOUT audio stream
The video display show the video stream from my IP camera.  The recording video button is enabled (red/round). Nothing happens when the record button is pressed, i.e. the button is unchanged (red/round). The video folder is empty afterwards.
Logfile: [QGC-v4.2.3-camera-h264-noaudio.txt](https://github.com/mavlink/qgroundcontrol/files/9814075/QGC-v4.2.3-camera-h264-noaudio.txt)

jensensi commented:
@fredowski 
Yes, I can tell that the logfiles in my [reply](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1279726539) @zdanek unfortunately are empty. I am not sure how that happend?? 
I have searched for the logfiles on my computer and on the device, but they have already been deleted.
I can reproduce the logfiles if needed?

I have posted similar logfiles (with content) in a [later post](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282846881) - including both lowercase 'rtsp' and 'WAITING FOR VIDEO'

fredowski commented:
Thanks again for your testing! This becomes now a little bit difficult to debug as I cannot reproduce the problem here. My guess is that this is again a problem of the dynamic addition of the decoder part. I moved the linking of the decoder with the valve to a probe idle callback as proposed here: https://gstconf.ubicast.tv/videos/how-to-work-dynamic-pipelines

Can you try the new build with your previous test?
Can you try also h.265?
Is there a difference if you use the first (main) stream according to your configuration screenshot?

jensensi commented:
@fredowski 
I will be happy to share the camera stream with you, so you can have a first-hand experience yourself :-)
Thank you for sharing the link to dynamic pipelines. I will study the information.

**The tests I have conducted with the 'new build'**
Build reference:  [https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) 
(Note, that I have called your 'new build' the 'Bunnybuild' in some comments above).

> Can you try the new build with your previous test?

The [previous test](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282846881) I did, was indeed with your 'new build' (and also with a standard QGC build - for reference).
In my previous test, I used a RTSP stream from my camera configured for H264 (not H265) - WITH and WITHOUT audio.

- Test cases 1.1) and 1.2) in [#issuecomment-1282846881](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282846881) use your ['new build'](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) with a H264 video stream from my camera.

> Can you try also h.265?

Yes, sure! I have updated my previous test of the 'new build' to also include H265 video stream from my camera - WITH and WITHOUT audio.
- The NEW test cases 3.1) and 3.2) in [#issuecomment-1282846881](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282846881) use your ['new build'](https://github.com/mavlink/qgroundcontrol/actions/runs/3273527680) with a H265 video stream from my camera.

> Is there a difference if you use the first (main) stream according to your configuration screenshot?

In all the test I have reported on this site, I have used a RTSP video stream configured for _either_ H264 or H265 - but _always_ using the 'mainprofile'. I have done some sporadic testing with other profiles initially (before I started to do logs), but I did not experience any difference in behavios at that time.

I will be happy to do further testing of the different profiles. The Bunny video is H264 'highprofile'.

jensensi commented:
@fredowski 
Today I have experimented with _gst-launch-1.0.exe_. The overall objective was to create a new complete recording 'pipeline' (in parallel to the current pipeline) in QGC for Windows 10 using 'gst_parse_launch', and then test if the same 'pipeline' also works on my Android device, i.e. does GStreamer recording work on my Android device at all??

I successfully added some code to QGC which enabled me to establish a new/parallel 'pipeline' that recorded the RTSP H265 video stream from my camera to a file in the QGC 'video directory' on Windows 10.
I then build the same code for Android to confirm if the code would also create a video file on my Android device? And it did :-)
I can also play back the video file afterwards (on both devices). 

I used the following 'pipeline' definition in the code:
> **gst-launch-1.0.exe rtspsrc name=myRtsp ! location=rtsp://192.168.0.130:554/12 ! application/x-rtp, stream-format=byte-stream ! rtph265depay ! h265parse ! matroskamux ! filesink name=fileSink location=test.mkv**

I have not tried a lot of different pipeline definitions, so there may still be room for improvement. But I was able to record video from my camera on my Android device, which convinces me that it will be possible to (somehow) do an RTSP H265 stream recording also with the QGC pipeline, so I will continue to search for solutions.

fredowski commented:
@jensensi congratulations to setup the gstreamer pipeline! I tested your stream and I can play it with

`gst-launch-1.0 rtspsrc location=rtsp://xxxx ! decodebin ! autovideosink`

The problems stem from the dynamic reconfiguration of the pipeline in QGroundControl. Here is an image that i created with the dot debugging from the running pipeline for the bunny rtsp movie:

![bild](https://user-images.githubusercontent.com/1280457/196880274-933ed63d-793a-4816-8ce2-f6b1d0709745.png)

You see the "parsebin" element before the "tee". The "tee" is used to distribute the stream to the video "decoder" branch that is currently active and the "recorder" branch which is not active. The idea is that you have a running rendered video. Then you push the red "record" button and then the recording branch will be added to the running pipeline. QGC now adds more complication by adding also the "decoding" branch dynamically. First the part before and including the "tee", the "queues" after tee and the "valves" are started. If a stream is detected, then the "decoder" branch is added and the video is rendered. These dynamic modifications account for much of the problems and the required code. In the 3.9.6 version of QGC the decoding was configured before starting the pipeline and only the recording was added later. There is also some description about the pipelines in the code. See: https://github.com/mavlink/qgroundcontrol/blob/v3.5.6/src/VideoStreaming/VideoReceiver.cc

Maybe it would be an idea to create the pipeline in the code also with gst_parse_launch and to step back from the addition of the decoding branch to the playing pipeline.

Comparing the qgc pipeline with your pipeline, qgc also uses "parsebin" and "decodebin" to let gstreamer automagically create the pipeline depending on the actual stream, i.e. instead of using "rtph265depay" and "h265parse" directly, this is selected by "parsebin" and "decodebin". 

With "new build" I mean the build where I added the last patch 
[Video/Gstreamer: Link Valve pad in IDLE time](https://github.com/mavlink/qgroundcontrol/pull/10420/commits/40bf94258ec480995c38b694ac4ce6a949fd25e2) to add the decoder in an idle callback. That is the Android build:  https://github.com/mavlink/qgroundcontrol/actions/runs/3281045089

jensensi commented:
@fredowski
I have now made the tests from my previous test (https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1282846881) using your latest build:
https://github.com/mavlink/qgroundcontrol/actions/runs/3281045089

**Testing**
I have tested the software on an RTSP stream from my IP Camera using both H264 and H265 encoding. In both cases I used the 'mainprofile' and the audio was disabled (no audio).

The logs are created using the settings:

- VideoAllLog = enabled
- GStreamer Debug = FixMe (GST_DEBUG=3)

**Findings**
Using software: [https://github.com/mavlink/qgroundcontrol/actions/runs/3281045089](https://github.com/mavlink/qgroundcontrol/actions/runs/3281045089)

_Test 1) Video source: IP camera, H.264 video stream, no audio_
The video display show "WAITING FOR VIDEO". No video is displayed. The recording video button is enabled (red/round).
When the record button is pressed, the button change shape from round to square (recording started), but it remains unchanged (square/red) when the button is pressed again. The video folder contain a new video file, which is empty (0 bytes).
[IP Camera-H264-noaudio.txt](https://github.com/mavlink/qgroundcontrol/files/9841584/IP.Camera-H264-noaudio.txt)


_Test 2) Video source: IP camera, H.265 video stream, no audio_
The video display show the video stream from my IP camera. The recording video button is enabled (red/round). Nothing happens when the record button is pressed, i.e. the button remains unchanged (red/round). The video folder is empty afterwards.
[IP Camera-H265-noaudio.txt](https://github.com/mavlink/qgroundcontrol/files/9841585/IP.Camera-H265-noaudio.txt)

fredowski commented:
I did some experiments with a Galaxy S5 mini android phone and a Macbook by changing GstVideoReceiver.cc with a completely cooked down version using gst_parse_launch.  

```
    GError *err = NULL;
    _pipeline = gst_parse_launch("uridecodebin uri=rtsp://wowzaec2demo.streamlock.net/vod/MP4:BigBuckBunny_115k.mp4 ! glupload ! glcolorconvert ! qmlglsink name=mysink sync=false", &err);
    GstElement *mysink = gst_bin_get_by_name(GST_BIN(_pipeline), "mysink");
    QQuickItem* widget;
    g_object_get(sink, "widget", &widget, NULL);
    g_object_set(mysink, "widget", widget, NULL);
    gst_element_set_state(_pipeline, GST_STATE_PLAYING);
    _dispatchSignal([this](){
        emit onStartDecodingComplete(STATUS_OK);
    });
    _dispatchSignal([this](){
        emit decodingChanged(true);
    });
    return; /**************************************************************************/
```

This code will simply start a pipeine with no other code interfering. The movie is a h264 coded rtsp stream including audio tracks. It turns out that:

  * uridecodebin shows the video on MacOS but not on Android
  * uridecodebin3 shows the video on Android but not non MacOS

decodebin and decodebin3 differ in the way the pipeline is "automagically" constructed. I also tested this with the stream from @jensensi. Further investigation showed:

**MacOS (with Bunnyvideo) **

decodebin tries the vtdec_hw apple hardware decoder. That decoder cannot convert and then decodebin switches to the software decoder avdec_h264 and the video is shown.

decodebin3 tries the vtdec_hw, it fails but decodebin3 does not switch to the avdec_h264 fallback. Video is not shown.

See fallback problem for decodebin3 here: https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/issues/468
[macos-decodebin-bunny.txt](https://github.com/mavlink/qgroundcontrol/files/9847295/macos-decodebin-bunny.txt)
[macos-decodebin3-bunny.txt](https://github.com/mavlink/qgroundcontrol/files/9847296/macos-decodebin3-bunny.txt)

**Android (with IP camera)**

decodebin selects the amcvideodec-omxexynosavcdec hardware decoder which negotiates a NV12 output format that the color converter downward the pipeline cannot convert. No video is shown.

decodebin3 also uses the hardware decoder but as a result of the downstream negotiation the decoder produces RGBA such that the color converter downstream converts RGBA to RGBA and the video is shown.

decodebin3 can better incorporate downstream requirements during format negotiation.
[android-decodebin.txt](https://github.com/mavlink/qgroundcontrol/files/9847298/android-decodebin.txt)
[android-decodebin3.txt](https://github.com/mavlink/qgroundcontrol/files/9847300/android-decodebin3.txt)

**Summary**
So decodebin3 does not work on MacOS and decodebin does not work on Android...

jensensi commented:
@fredowski

Great work!! Very structured and analytic - and with some arrently very specific findings. 
Unfortunately, I do not have sufficient gstreamer skills to conclude anything from you findings - except perhaps that it is not possible to satisfy all stream encodings on all platforms with one pipeline as decodebin & decodebin3 behave differently on different platforms.
Is the finding affecting both the displaying and recording of an rtsp stream?
What is the required approach? Gstreamer correction/improvement? Updating QGC implementation adapt to the platform e.g at build time?

I wanted to reproduce your pipeline on my Android device in order to contribute with similar logging files taken from my device, but I am unable to get the pipeline to work?
I have experiemented with the pipeline string using gst-launch-1.0.exe on my Windows 10 PC without luck.
So I do not know how to contribute for this specific pipeline?

What GStreamer debug level have you used for the log files in [this last test](https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1288175173)? Is it FixMe (GST_DEBUG=3)?

Also, I will try to see if I can get the dot debugging tool up and running from my Android device.

Updates to follow.

fredowski commented:
I added a patch thats shows the usage of gst_parse_launch. This only renders the video with no recording. Just to show how it works. I think the way to go is decodebin3. If this does not work for certain streams on MacOS like the bunny stream or 320x240 h264, then it is a bug in gstreamer which should be fixed there. I tested this on MacOS with vlc for mpeg/ts streams, your camera for rtsp and gstreamer commandline for raw h264 streams.

andrewvoznytsa commented:
> If this does not work for certain streams on MacOS like the bunny stream or 320x240 h264, then it is a bug in gstreamer which should be fixed there.

I don't think there is any bug on gstreamer side. From what I recall the problem is media type (GstCaps) produced by source element (its src pad) and this one supported on sink pad of decoder element. If I'm not mistaken then some source elements send H.264 frames as NALU aligned byte stream while others do generate so-called H.264 AU AVC samples. Some Android and iOS/MacOS HW decoders require H.264 NALU, some - H.264 AVC AU. Using avdec_h264 which is ffmpeg's software decoder and behaves identically on all system you should be able to decode on all platforms (with limited performance sometimes).

Your best friend is gst-inspect-1.0 <element name> to see supported caps and then you can think how to make your pipeline working on all platforms. Making this in automatic way is the worst possible solution, IMHO. It will never work good enough on all systems. Doing that manually is also not the best solution but at least it will work according to your requirements (it is what I did a few years ago).

There is no silver bullet - just some amount of information to learn and then not-so-big task to complete.

H.265 issue is close to that - make sure that the same format is used on all ends and all components support it. Sometimes some pads should be linked dynamically, when it is fully initialized instead of this kind of 'static' linking which I used before.

andrewvoznytsa commented:
Converting in between H.264 NALU and H.264 AVC AU is also not a rocket science - h264parse does that. It might add some extra latency (in some cases it could be quite high, on average it should not be more than +1 frame to your existing latency).

In certain cases elements that produces H.264 frame supports both formats. Then capsfilter is your friend - you can get this format that is best current case with zero extra latency.

The trick is to make it working on all platforms, keep control on everything and at the same time don't spend a lot of time maintaining manual pipeline building.

fredowski commented:
@andrewvoznytsa : Thanks for looking into this! As you say: "The trick is to make it work on all platforms with as general as possible pipeline construction." The bug description for the failing fallback for decodebin3 is here:  https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/issues/468.

jensensi commented:
@andrewvoznytsa :
During initial testing ([#10366](https://github.com/mavlink/qgroundcontrol/issues/10366#issuecomment-1220475094)) of QGC built for different platforms (WIndows10 & Android) with the one and same IP camera, streaming in H265 on RTSP, I found that the caps of the 'src pad' changed between running on Windows 10 and Android??

- Windows 10:   stream-format = 'hvc1'
- Android: stream-format = 'byte-stream'

It really surprised me!?
Is this behavior related to your statement:

>  From what I recall the problem is media type (GstCaps) produced by source element (its src pad) and this one supported on sink pad of decoder element. If I'm not mistaken then some source elements send H.264 frames as NALU aligned byte stream while others do generate so-called H.264 AU AVC samples.

Davidsastresas commented:
I tested the 6 commits presented in this PR on an android poco X3 nfc, a skydroid h16, and a samsung galaxy s8, but in all of them the video recording was still not working. I added them to QGC master. Should it be working?

I am happy to keep testing whatever is needed. Thanks!

fredowski commented:
@Davidsastresas thanks for testing! With the last commit this is only a technlogy test to use gst_launch_parse. That is a result of the discussion https://github.com/mavlink/qgroundcontrol/pull/10420#issuecomment-1288175173. The recording is known to be broken in this test. 

Some further investigation showed the the following pipeline

```gst-launch-1.0 -vvvv udpsrc address=127.0.0.1 port=5600 ! parsebin ! queue ! decodebin3 ! autovideosink```

which is the basic structure in QGC without the tees and valves cannot display a mpeg-ts h264 video played from vlc player on MacOS, while 

  * ```udpsrc address=127.0.0.1 port=5600  ! decodebin3 ! autovideosink```
  * ```udpsrc address=127.0.0.1 port=5600 ! parsebin ! queue ! decodebin ! autovideosink```
  * ```udpsrc address=127.0.0.1 port=5600 ! decodebin ! autovideosink```

all work and show the video. QGC 4.2.3 also does not show the video. This is all pretty much a mess.

julianoes commented:
I'm just going through PRs and stumbled on this. Is it still applicable?

fredowski commented:
My conclusion is that this not fixable on all platforms. I just close this.

jensensi commented:
A few contributing comments regarding my further experiences and findings.
At the time of my reports, I was in dire need of the recording functionality, so I ended up defining a separate button and pipeline to overcome the challenges I was facing, using the following pipeline definition with success:

```
QString pipeline_definition = QString("-e rtspsrc location=%1 ! rtph265depay ! h265parse ! matroskamux ! filesink name=fileSink").arg(url);
_pipeline= gst_parse_launch(qPrintable(pipeline_definition), NULL);
_sink = gst_bin_get_by_name((GstBin*)_pipeline,"fileSink");
```

I have not used/tested the QGC standard recording functionality since.
Should I get a chance to test the recording feature on a more recent QGC release and it fails, then I will just open a new report.
